chalk-recursive/src/recursive.rs

```diff
@@ -224,20 +239,40 @@ impl<'me, I: Interner> SolveDatabase<I> for Solver<'me, I> {
             // Check if this table is still on the stack.
             if let Some(depth) = self.context.search_graph[dfn].stack_depth {
                 // Is this a coinductive goal? If so, that is success,
-                // so we can return normally. Note that this return is
-                // not tabled.
-                //
-                // XXX how does caching with coinduction work?
+                // so we can return and set the minimum to its DFN.
+                // Note that this return is not tabled. And so are
+                // all other solutions in the cycle until the cycle
+                // start is finished. This avoids prematurely cached
+                // false positives.
                 if self.context.stack.coinductive_cycle_from(depth) {
-                    let value = ConstrainedSubst {
-                        subst: goal.trivial_substitution(self.program.interner()),
-                        constraints: Constraints::empty(self.program.interner()),
-                    };
-                    debug!("applying coinductive semantics");
-                    return Ok(Solution::Unique(Canonical {
-                        value,
-                        binders: goal.canonical.binders,
-                    }));
+                    if self.context.search_graph[dfn].coinductive_start {
+                        // The cycle start was already marked as such
+                        // and its result can be returned.
+                        minimums.update_from(self.context.search_graph[dfn].links);
+                        return self.context.search_graph[dfn].solution.clone();
+                    } else {
+                        // The start of the coinductive cycle
+                        // has not been marked as such yet.
+                        // Assume a positive result for now.
+
+                        let value = ConstrainedSubst {
+                            subst: goal.trivial_substitution(self.program.interner()),
+                            constraints: Constraints::empty(self.program.interner()),
+                        };
+                        let solution = Ok(Solution::Unique(Canonical {
+                            value,
+                            binders: goal.canonical.binders,
+                        }));
+
+                        debug!("applying coinductive semantics");
+                        minimums.update_from(self.context.search_graph[dfn].links);
+
+                        // Mark the start of the coinductive cycle and assume the positive result.
+                        self.context.search_graph[dfn].coinductive_start = true;
+                        self.context.search_graph[dfn].solution = solution.clone();
+
+                        return solution;
+                    }
                 }
 
                 self.context.stack[depth].flag_cycle();
```

Comment:
It's not part of the diff, but I think we want to rewrite this code:

https://github.com/rust-lang/chalk/blob/a0eac73f60ad0ab1de2075def346bd5a5f2352cd/chalk-recursive/src/recursive.rs#L257-L262

so that the `self.context.search_graph.insert` doesn't always set the "initial result" to `Err`. Instead, when `coinductive_goal` is true, the initial result should be `Ok()` with an identity substitution.

Reply:
I'd suspect that this introduces a bit of unnecessary overhead for some goals but it also allows to greatly simplify most of the code. I'm not sure whether and if how performance could be tested for this.

Reply:
Why would it introduce overhead?

Reply:
The generation of a solution with identity substitution is more computation expansive than the `Err`. Necessary is this start value only for the start goals of coinductive cycles as far as I understand. Thus, the changes might introduce unnecessary overhead for some goals. On the other hand, the explicit handling of coinductive cycles would probably introduce even more overhead so I would expect no problem in that regard.

Reply:
I see. I think it's worth it. We could specialize the representation of identity substitution if needed (e.g., make an enum like `enum Solution { Identity, Subst(...) }`